execution/function/functions.go

```diff
@@ -455,9 +492,62 @@ func extrapolatedRate(samples []promql.Point, isCounter, isRate bool, stepTime i
 	if isRate {
 		factor /= float64(selectRange / 1000)
 	}
-	resultValue *= factor
+	if resultHistogram == nil {
+		resultValue *= factor
+	} else {
+		resultHistogram.Scale(factor)
+
+	}
+
+	return resultValue, resultHistogram
+}
+
+// histogramRate is a helper function for extrapolatedRate. It requires
+// points[0] to be a histogram. It returns nil if any other Point in points is
+// not a histogram.
+func histogramRate(points []promql.Point, isCounter bool) *histogram.FloatHistogram {
+	prev := points[0].H // We already know that this is a histogram.
+	last := points[len(points)-1].H
+	if last == nil {
+		return nil // Range contains a mix of histograms and floats.
+	}
+	minSchema := prev.Schema
+	if last.Schema < minSchema {
+		minSchema = last.Schema
+	}
 
-	return resultValue
+	// First iteration to find out two things:
+	// - What's the smallest relevant schema?
+	// - Are all data points histograms?
+	//   TODO(beorn7): Find a way to check that earlier, e.g. by handing in a
```

Comment:
Let's remove the TODO here, or at least replace the github handle with someone else. There are some other TODOs in this PR that should be either removed or adjusted.